### PR TITLE
feat: add support for optimistic concurrency check

### PR DESCRIPTION
**Pull request by migmartri**
This is the first step (API, biz, data layers) to enable optimistic concurrency in a remote state. It contains:

- The state API now returns the digest of the current state.
- This digest needs to be sent back during an update; if it mismatches with the content stored, it will return a known conflict error.
- This new error will be handled in another PR by the CLI, which will retry accordingly.

This digest check happens in a DB transaction now, including the use of `SELECT FOR UPDATE`, which required enabling locking in `ent`, which is the reason of the `data/ent` changes in this PR. 

Example of read response.

image

Write also supports an optional `baseDigest` that will be used for checking

ref #973

### CODE REVIEW

File: app/controlplane/api/controlplane/v1/attestation_state.proto [hunk 1]
```diff
@@ -17,18 +17,19 @@ syntax = "proto3";
 
 package controlplane.v1;
 
-option go_package = "github.com/chainloop-dev/chainloop/app/controlplane/api/controlplane/v1;v1";
-
-import "buf/validate/validate.proto";
 import "attestation/v1/crafting_state.proto";
+import "buf/validate/validate.proto";
```
Comment (migmartri) [inline app/controlplane/api/controlplane/v1/attestation_state.proto:21]: buf format seems to make these changes now, I am not sure why

File: app/controlplane/pkg/data/attestationstate.go [hunk 2]
```diff
@@ -48,26 +51,69 @@ func (r *AttestationStateRepo) Initialized(ctx context.Context, runID uuid.UUID)
 	return exists, nil
 }
 
-func (r *AttestationStateRepo) Save(ctx context.Context, runID uuid.UUID, state []byte) error {
-	err := r.data.DB.WorkflowRun.UpdateOneID(runID).SetAttestationState(state).Exec(ctx)
+// baseDigest, when provided will be used to check that it matches the digest of the state currently in the DB
+// if the digests do not match, the state has been modified and the caller should retry
+func (r *AttestationStateRepo) Save(ctx context.Context, runID uuid.UUID, state []byte, baseDigest string) (err error) {
```
Comment (migmartri) [inline app/controlplane/pkg/data/attestationstate.go:56]: this method has the meat of the implementation, it makes sure that 

in a transaction

- Reads the state stored in DB and blocks the row for update
- Calculates the sha of its content and compares it with the baseDigest optionally provided
- if it doesn't match, it returns a ConflictError and cancels the transaction, unblocking the row.
- If it does, we update the column and unblock the row

File: app/controlplane/api/controlplane/v1/attestation_state.proto [hunk 3]
```diff
@@ -68,3 +74,8 @@ message AttestationStateServiceResetRequest {
 
 message AttestationStateServiceResetResponse {}
 
+enum AttestationStateError {
+  ATTESTATION_STATE_ERROR_UNSPECIFIED = 0;
+  // https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409
+  ATTESTATION_STATE_ERROR_CONFLICT = 1 [(errors.code) = 409];
```
Comment (jiparis) [inline app/controlplane/api/controlplane/v1/attestation_state.proto:80]: Nice! we can set error codes in the spec!

File: app/controlplane/pkg/data/attestationstate.go [hunk 4]
```diff
@@ -48,26 +51,69 @@ func (r *AttestationStateRepo) Initialized(ctx context.Context, runID uuid.UUID)
...
+
+	defer func() {
+		// Unblock the row if there was an error
+		if err != nil {
+			_ = tx.Rollback()
+		}
+	}()
+
+	// compared the provided digest with the digest of the state in the DB
+	// TODO: make digest check mandatory on updates
```
Comment (jiparis) [inline app/controlplane/pkg/data/attestationstate.go:70]: I guess we can do this when all clients have been updated, right?

**Review by jiparis — APPROVED** [review]

File: app/controlplane/pkg/data/attestationstate.go (hunk 4 above)
Comment (migmartri) [reply app/controlplane/pkg/data/attestationstate.go:70]: Correct, for now it's not required.

**Pull request merged**